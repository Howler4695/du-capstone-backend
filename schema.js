// Pre-seeded data
const books = [
  {
    id: '1',
    title: 'Harry Potter and the Chamber of Secrets',
    author: '1',
    coverImage:
      'https://m.media-amazon.com/images/I/51mFoFmu0EL._AC_SY780_.jpg',
    categories: ['1', '2'],
    description:
      'Harry Potter and the Chamber of Secrets is a 1998 young adult fantasy novel by J.K. Rowling, the second in the Harry Potter series. The story follows Harry’s tumultuous second year at Hogwarts School of Witchcraft and Wizardry, including an encounter with Voldemort, the wizard who killed Harry’s parents. Against this fantastic backdrop, Rowling examines such themes as death, fame, friendship, choice, and prejudice. Upon release, the novel became a worldwide bestseller and won several awards, including Children’s Book of the Year at the British Book Awards and the Nestlé Smarties Book Award; it was subsequently adapted into a 2002 film directed by Chris Columbus.'
  },
  {
    id: '2',
    title: 'Harry Potter and the Prisoner of Azkaban',
    author: '1',
    coverImage:
      'https://m.media-amazon.com/images/I/51DQeuJ5QDL._SY291_BO1,204,203,200_QL40_FMwebp_.jpg',
    categories: ['1', '2']
  },
  {
    id: '3',
    title: 'Harry Potter and the Goblet of Fire',
    author: '1',
    coverImage:
      'https://m.media-amazon.com/images/I/51gy+g8Z+1L._SX343_BO1,204,203,200_.jpg',
    categories: ['1', '2']
  },
  {
    id: '4',
    title: 'C All in One Desk Reference For Dummies',
    author: '2',
    coverImage:
      'https://m.media-amazon.com/images/I/51LNAmfIQ3L._SX404_BO1,204,203,200_.jpg',
    categories: ['3']
  }
];
const authors = [
  {
    id: '1',
    firstName: 'J.K.',
    lastName: 'Rowling',
    books: ['1', '2', '3']
  },
  {
    id: '2',
    firstName: 'Dan',
    lastName: 'Gookin',
    books: ['4']
  },
  {
    id: '3',
    firstName: 'Ayn',
    lastName: 'Rand',
    books: []
  },
  {
    id: '4',
    firstName: 'George',
    lastName: 'Orwell',
    books: []
  },
  {
    id: '5',
    firstName: 'John',
    lastName: 'Steinbeck',
    books: []
  },
  {
    id: '6',
    firstName: 'F. Scott',
    lastName: 'Fitzgerald',
    books: []
  },
  {
    id: '7',
    firstName: 'Ray',
    lastName: 'Bradbury',
    books: []
  },
  {
    id: '8',
    firstName: 'William',
    lastName: 'Faulkner',
    books: []
  }
];
const categories = [
  {
    id: '1',
    name: 'Fantasy',
    books: ['1', '2', '3']
  },
  {
    id: '2',
    name: 'Fiction',
    books: ['1', '2', '3']
  },
  {
    id: '3',
    name: 'Programming',
    books: ['4']
  }
];
