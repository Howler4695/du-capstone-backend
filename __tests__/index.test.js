import { books, categories } from '../data/schema.js';
import newTestServer from './utils/test-server.js';

describe('index', () => {
  const testServer = newTestServer();
  const originalBooks = [...books];

  describe('getBooks', () => {
    const query = `
      query {
        getBooks {
          id
          title
          author {
            id
          }
          coverImage
          categories {
            id
          }
          description
        }
      }
    `;

    it('should return all books', async () => {
      const { body: response } = await testServer.executeOperation({ query });
      expect(response).toMatchSnapshot();
    });
  });

  describe('getAuthors', () => {
    const query = `
      query {
        getAuthors {
          id
          firstName
          lastName
          books {
            id
          }
        }
      }
    `;

    it('should return all authors', async () => {
      const { body: response } = await testServer.executeOperation({ query });
      expect(response).toMatchSnapshot();
    });
  });

  describe('getCategories', () => {
    const query = `
      query {
        getCategories {
          id
          name
          books {
            id
          }
        }
      }
    `;

    it('should return all categories', async () => {
      const { body: response } = await testServer.executeOperation({ query });
      expect(response).toMatchSnapshot();
    });

    describe('getBooksByIds', () => {
      const query = `
      query GetBooksByIds($bookIds: [ID!]) {
        getBooksByIds(bookIds: $bookIds) {
          title
        }
      }
    `;

      it('should return books with corresponding ids', async () => {
        const queryIds = ['1', '2', '4'];

        const { body: response } = await testServer.executeOperation({
          query,
          variables: { bookIds: queryIds }
        });
        const { data, errors } = response.singleResult;

        expect(data.getBooksByIds).toBeDefined();
        expect(errors).toBeUndefined();

        const { getBooksByIds: testBookTitles } = data;

        expect(testBookTitles.map(book => book.title)).toEqual([
          'Harry Potter and the Chamber of Secrets',
          'Harry Potter and the Prisoner of Azkaban',
          'C All in One Desk Reference For Dummies'
        ]);
      });

      it('should return books with corresponding ids, and no duplicates', async () => {
        const queryIds = ['1', '2', '2'];

        const { body: response } = await testServer.executeOperation({
          query,
          variables: { bookIds: queryIds }
        });
        const { data, errors } = response.singleResult;

        expect(data.getBooksByIds).toBeDefined();
        expect(errors).toBeUndefined();

        const { getBooksByIds: testBookTitles } = data;

        expect(testBookTitles.map(book => book.title)).toEqual([
          'Harry Potter and the Chamber of Secrets',
          'Harry Potter and the Prisoner of Azkaban'
        ]);
      });
    });

    describe('addBook', () => {
      const mutation = `
      mutation AddBook($newBook: BookInput!) {
        addBook(newBook: $newBook) {
          id
          title
          author {
            id
          }
          coverImage
          categories {
            id
          }
          description
        }
      }
    `;

      afterEach(() => {
        books.pop();
      });

      it('should add book to memory and return book', async () => {
        const newBook = {
          title: 'Starship Troopers',
          author: '2',
          coverImage: 'https://incredible-cover-image.jpeg',
          categories: ['1', '2'],
          description: "It's starship troopers"
        };

        const { body: response } = await testServer.executeOperation({
          query: mutation,
          variables: { newBook }
        });
        const { data, errors } = response.singleResult;

        expect(data.addBook).toBeDefined();
        expect(errors).toBeUndefined();

        const testBook = data.addBook;

        expect(testBook.id).toBe('5');
        expect(testBook.title).toBe('Starship Troopers');
        expect(testBook.author.id).toBe('2');
        expect(testBook.coverImage).toBe('https://incredible-cover-image.jpeg');
        expect(testBook.categories.map(category => category.id)).toEqual([
          '1',
          '2'
        ]);
        expect(testBook.description).toBe("It's starship troopers");
      });
    });
  });

  describe('addCategory', () => {
    const mutation = `
    mutation AddCategory($newCategory: CategoryInput!) {
      addCategory(newCategory: $newCategory) {
        id
        name
        books {
          id
        }
      }
    }
  `;

    afterEach(() => {
      categories.pop();
    });

    it('should add category to memory and return category', async () => {
      const newCategory = {
        name: 'Religous',
        books: ['2', '3']
      };

      const { body: response } = await testServer.executeOperation({
        query: mutation,
        variables: { newCategory }
      });
      const { data, errors } = response.singleResult;

      expect(data.addCategory).toBeDefined();
      expect(errors).toBeUndefined();

      const testCategory = data.addCategory;

      expect(testCategory.id).toBe('4');
      expect(testCategory.name).toBe('Religous');
      expect(testCategory.books.map(book => book.id)).toEqual(['2', '3']);
    });
  });

  describe('updateBook', () => {
    const mutation = `
    mutation UpdateBook($bookId: ID!, $updatedBook: BookInput!) {
      updateBook(bookId: $bookId, updatedBook: $updatedBook) {
        id
        title
        author {
          id
        }
        coverImage
        categories {
          id
        }
        description
      }
    }
  `;

    afterEach(() => {
      books[1] = originalBooks[1];
    });

    it('should update book', async () => {
      const bookId = '2';
      const updatedBook = {
        title: 'Starship Troopers',
        author: '2',
        coverImage: 'https://incredible-cover-image.jpeg',
        categories: ['1', '2'],
        description: "It's starship troopers"
      };

      const { body: response } = await testServer.executeOperation({
        query: mutation,
        variables: { bookId, updatedBook }
      });
      const { data, errors } = response.singleResult;

      expect(data).toBeDefined();
      expect(errors).toBeUndefined();

      const testBook = data.updateBook;

      expect(testBook.id).toBe('2');
      expect(testBook.title).toBe('Starship Troopers');
      expect(testBook.author.id).toBe('2');
      expect(testBook.coverImage).toBe('https://incredible-cover-image.jpeg');
      expect(testBook.categories.map(category => category.id)).toEqual([
        '1',
        '2'
      ]);
      expect(testBook.description).toBe("It's starship troopers");
    });
  });
});
