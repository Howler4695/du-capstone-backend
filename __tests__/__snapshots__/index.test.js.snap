// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index getAuthors should return all authors 1`] = `
{
  "kind": "single",
  "singleResult": {
    "data": {
      "getAuthors": [
        {
          "books": [
            {
              "id": "1",
            },
            {
              "id": "2",
            },
            {
              "id": "3",
            },
          ],
          "firstName": "J.K.",
          "id": "1",
          "lastName": "Rowling",
        },
        {
          "books": [
            {
              "id": "4",
            },
          ],
          "firstName": "Dan",
          "id": "2",
          "lastName": "Gookin",
        },
        {
          "books": [],
          "firstName": "Ayn",
          "id": "3",
          "lastName": "Rand",
        },
        {
          "books": [],
          "firstName": "George",
          "id": "4",
          "lastName": "Orwell",
        },
        {
          "books": [],
          "firstName": "John",
          "id": "5",
          "lastName": "Steinbeck",
        },
        {
          "books": [],
          "firstName": "F. Scott",
          "id": "6",
          "lastName": "Fitzgerald",
        },
        {
          "books": [],
          "firstName": "Ray",
          "id": "7",
          "lastName": "Bradbury",
        },
        {
          "books": [],
          "firstName": "William",
          "id": "8",
          "lastName": "Faulkner",
        },
      ],
    },
    "errors": undefined,
  },
}
`;

exports[`index getBooks should return all books 1`] = `
{
  "kind": "single",
  "singleResult": {
    "data": {
      "getBooks": [
        {
          "author": {
            "id": "1",
          },
          "categories": [
            {
              "id": "1",
            },
            {
              "id": "2",
            },
          ],
          "coverImage": "https://m.media-amazon.com/images/I/51mFoFmu0EL._AC_SY780_.jpg",
          "description": "Harry Potter and the Chamber of Secrets is a 1998 young adult fantasy novel by J.K. Rowling, the second in the Harry Potter series. The story follows Harry’s tumultuous second year at Hogwarts School of Witchcraft and Wizardry, including an encounter with Voldemort, the wizard who killed Harry’s parents. Against this fantastic backdrop, Rowling examines such themes as death, fame, friendship, choice, and prejudice. Upon release, the novel became a worldwide bestseller and won several awards, including Children’s Book of the Year at the British Book Awards and the Nestlé Smarties Book Award; it was subsequently adapted into a 2002 film directed by Chris Columbus.",
          "id": "1",
          "title": "Harry Potter and the Chamber of Secrets",
        },
        {
          "author": {
            "id": "1",
          },
          "categories": [
            {
              "id": "1",
            },
            {
              "id": "2",
            },
          ],
          "coverImage": "https://m.media-amazon.com/images/I/51DQeuJ5QDL._SY291_BO1,204,203,200_QL40_FMwebp_.jpg",
          "description": null,
          "id": "2",
          "title": "Harry Potter and the Prisoner of Azkaban",
        },
        {
          "author": {
            "id": "1",
          },
          "categories": [
            {
              "id": "1",
            },
            {
              "id": "2",
            },
          ],
          "coverImage": "https://m.media-amazon.com/images/I/51gy+g8Z+1L._SX343_BO1,204,203,200_.jpg",
          "description": null,
          "id": "3",
          "title": "Harry Potter and the Goblet of Fire",
        },
        {
          "author": {
            "id": "2",
          },
          "categories": [
            {
              "id": "3",
            },
          ],
          "coverImage": "https://m.media-amazon.com/images/I/51LNAmfIQ3L._SX404_BO1,204,203,200_.jpg",
          "description": null,
          "id": "4",
          "title": "C All in One Desk Reference For Dummies",
        },
      ],
    },
    "errors": undefined,
  },
}
`;

exports[`index getCategories should return all categories 1`] = `
{
  "kind": "single",
  "singleResult": {
    "data": {
      "getCategories": [
        {
          "books": [
            {
              "id": "1",
            },
            {
              "id": "2",
            },
            {
              "id": "3",
            },
          ],
          "id": "1",
          "name": "Fantasy",
        },
        {
          "books": [
            {
              "id": "1",
            },
            {
              "id": "2",
            },
            {
              "id": "3",
            },
          ],
          "id": "2",
          "name": "Fiction",
        },
        {
          "books": [
            {
              "id": "4",
            },
          ],
          "id": "3",
          "name": "Programming",
        },
      ],
    },
    "errors": undefined,
  },
}
`;
